From foo@baz Tue Apr  9 12:12:43 2002
Date: Thu,  1 Nov 2007 09:29:06 -0600 (MDT)
To: Greg KH <greg@kroah.com>
From: Greg Kroah-Hartman <gregkh@suse.de>
Subject: kset: convert /sys/hypervisor to use kset_create

Dynamically create the kset instead of declaring it statically.  We also
rename hypervisor_subsys to hypervisor_kset to catch all users of the
variable.

Cc: Kay Sievers <kay.sievers@vrfy.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 arch/s390/hypfs/inode.c   |    2 +-
 drivers/base/hypervisor.c |   14 ++++++++++----
 include/linux/kobject.h   |    4 ++--
 3 files changed, 13 insertions(+), 7 deletions(-)

--- a/arch/s390/hypfs/inode.c
+++ b/arch/s390/hypfs/inode.c
@@ -506,7 +506,7 @@ static int __init hypfs_init(void)
 			goto fail_diag;
 		}
 	}
-	s390_subsys.kobj.kset = &hypervisor_subsys;
+	s390_subsys.kobj.kset = hypervisor_kset;
 	rc = subsystem_register(&s390_subsys);
 	if (rc)
 		goto fail_sysfs;
--- a/drivers/base/hypervisor.c
+++ b/drivers/base/hypervisor.c
@@ -2,19 +2,25 @@
  * hypervisor.c - /sys/hypervisor subsystem.
  *
  * Copyright (C) IBM Corp. 2006
+ * Copyright (C) 2007 Greg Kroah-Hartman <gregkh@suse.de>
+ * Copyright (C) 2007 Novell Inc.
  *
  * This file is released under the GPLv2
  */
 
 #include <linux/kobject.h>
 #include <linux/device.h>
-
+#include <linux/err.h>
 #include "base.h"
 
-decl_subsys(hypervisor, NULL);
-EXPORT_SYMBOL_GPL(hypervisor_subsys);
+struct kset *hypervisor_kset;
+EXPORT_SYMBOL_GPL(hypervisor_kset);
 
 int __init hypervisor_init(void)
 {
-	return subsystem_register(&hypervisor_subsys);
+	hypervisor_kset = kset_create_and_register("hypervisor", NULL,
+						   NULL, NULL);
+	if (IS_ERR(hypervisor_kset))
+		return PTR_ERR(hypervisor_kset);
+	return 0;
 }
--- a/include/linux/kobject.h
+++ b/include/linux/kobject.h
@@ -193,8 +193,8 @@ struct kset _name##_subsys = { \
 
 /* The global /sys/kernel/ kset for people to chain off of */
 extern struct kset *kernel_kset;
-/* The global /sys/hypervisor/ subsystem  */
-extern struct kset hypervisor_subsys;
+/* The global /sys/hypervisor/ kset for people to chain off of */
+extern struct kset *hypervisor_kset;
 
 extern int __must_check subsystem_register(struct kset *);
 extern void subsystem_unregister(struct kset *);
